tests/test_demography.py

```diff
@@ -570,6 +571,208 @@ def test_two_pop_change_growth_rates(self):
+                                                   population_id=0, growth_rate=7e-4)
+            ],
+            migration_matrix=[
+                [0, 0],
+                [0, 0]
+            ])
+        steps = np.linspace(0, 400, 401)
+        rates, P = ddb.coalescence_rate_trajectory(steps=steps, num_samples=[2, 0])
+        pop_sizes = ddb.population_size_trajectory(steps=steps)
+        for r, p in zip(rates, pop_sizes[:, 0]):
+            self.assertAlmostEqual(1/(2*r), p)
+        rates, P = ddb.coalescence_rate_trajectory(steps=steps, num_samples=[0, 2])
+        pop_sizes = ddb.population_size_trajectory(steps=steps)
+        for r, p in zip(rates, pop_sizes[:, 1]):
+            self.assertAlmostEqual(1/(2*r), p)
+
+    def test_rate_size_equality_with_population_merge(self):
+        # Test equality between two population that split with continued
+        # migration
+        N_A = 1e7
+        N_B = 1e6
+
+        ddb = msprime.DemographyDebugger(
+            population_configurations=[
+                msprime.PopulationConfiguration(initial_size=N_A),
+                msprime.PopulationConfiguration(initial_size=N_B)
+            ],
+            demographic_events=[
+                msprime.PopulationParametersChange(time=100, initial_size=1e5,
+                                                   population_id=0, growth_rate=0),
+                msprime.PopulationParametersChange(time=100, initial_size=1e5,
+                                                   population_id=1, growth_rate=7e-4),
+                msprime.MassMigration(time=100, source=1, dest=0),
+                msprime.PopulationParametersChange(time=200, initial_size=1e6,
+                                                   population_id=0, growth_rate=-1e-2),
+                msprime.PopulationParametersChange(time=300, initial_size=1e6,
+                                                   population_id=0, growth_rate=0),
+            ],
+            migration_matrix=[
+                [0, 0],
+                [0, 0]
+            ])
+        steps = np.linspace(0, 400, 401)
+        rates, P = ddb.coalescence_rate_trajectory(steps=steps, num_samples=[2, 0])
+        pop_sizes = ddb.population_size_trajectory(steps=steps)
+        for r, p in zip(rates[301:], pop_sizes[:, 0][301:]):
+            self.assertAlmostEqual(1/(2*r), p)
+
+
+class TestMatrixExponential(unittest.TestCase):
+    """
+    Test cases for the matrix exponential function.
+    """
+    def verify(self, A):
+        E1 = scipy.linalg.expm(A)
+        E2 = msprime.matrix_exponential(A)
+        self.assertEqual(E1.shape, E2.shape)
+        self.assertTrue(np.allclose(E1, E2))
+
+    def test_singleton(self):
+        for j in range(10):
+            A = np.array([[j]])
+            self.verify(A)
+
+    def test_zeros(self):
+        for j in range(1, 10):
+            A = np.zeros((j, j))
+            self.verify(A)
+
+    def test_ones(self):
```

Comment:
This is a stupid test (and is causing problems) - can we replace it with a few example matrices where we expect to do well?

Reply:
These are almost good tests; if these matrices are modified to be valid generators then we're good.